pkg/kube/client/ssa/caching_unstructured_extractor_test.go

```diff
@@ -0,0 +1,937 @@
+		"/openapi/v3/apis/pkg.crossplane.io/v1beta1":  "pkg.crossplane.io.v1beta1.json",
+		"/openapi/v3/apis/pkg.crossplane.io/v1":       "pkg.crossplane.io.v1.json",
+		"/openapi/v3/apis/nop.example.org/v1alpha1":   "nop.example.org.v1alpha1.json",
+	}
+	respFile, ok := responses[path]
+	if !ok {
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+	file, err := os.Open("test/openapi_schemas/" + respFile)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	io.Copy(w, file)
+}
+
+type mockAPIServer struct {
+	server        *httptest.Server
+	discoveryData handler3.OpenAPIV3Discovery
+}
+
+var discoveryDataInitial = map[string]string{
+	"api/v1":                          "111",
+	"apis/apps/v1":                    "111",
+	"apis/pkg.crossplane.io/v1alpha1": "111",
+	"apis/pkg.crossplane.io/v1beta1":  "111",
+	"apis/pkg.crossplane.io/v1":       "",
+	"apis/nop.example.org/v1alpha1":   "111",
+}
+
+var discoveryDataChangedRemoval = map[string]string{
+	"api/v1":                        "111",
+	"apis/apps/v1":                  "111",
+	"apis/pkg.crossplane.io/v1":     "",
+	"apis/nop.example.org/v1alpha1": "222",
+}
+
+func newMockAPIServer() (*mockAPIServer, error) {
+	as := &mockAPIServer{}
+
+	as.setDiscoveryData(discoveryDataInitial)
+
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch {
+		case strings.HasPrefix(r.URL.Path, "/openapi/v3/api"):
+			mockServeOAPISchema(r.URL.Path, w)
+
+		case r.URL.Path == "/openapi/v3":
+			// return root content
+			w.Header().Set("Content-Type", "application/json")
+			w.WriteHeader(http.StatusOK)
+			json.NewEncoder(w).Encode(as.discoveryData)
+		default:
+			w.WriteHeader(http.StatusNotFound)
+			return
+		}
+	}))
+	as.server = server
+	return as, nil
+}
+
+func (mas *mockAPIServer) setDiscoveryData(pathToHash map[string]string) {
+	disco := handler3.OpenAPIV3Discovery{
+		Paths: map[string]handler3.OpenAPIV3DiscoveryGroupVersion{},
+	}
+	for path, hash := range pathToHash {
+		hashParamStr := ""
+		if hash != "" {
+			hashParamStr = "?hash=" + hash
+		}
+		disco.Paths[path] = handler3.OpenAPIV3DiscoveryGroupVersion{
+			ServerRelativeURL: "/openapi/v3/" + path + hashParamStr,
+		}
+	}
+	mas.discoveryData = disco
+}
+
+func TestExtract(t *testing.T) {
+	tests := []struct {
+		args args
+		want want
+		name string
+	}{
+		{
+			name: "SuccessfulExtract",
+			args: args{
+				objectToExtract: []byte(testObjectForExtraction1),
+				fieldManager:    "dude",
+			},
+			want: want{
+				extractedObject: []byte(want1),
+			},
+		},
+		{
+			name: "SuccessfulExtractWithFieldManagerOwnsNothing",
+			args: args{
+				objectToExtract: []byte(testObjectForExtraction2),
+				fieldManager:    "another-guy",
+			},
+			want: want{
+				extractedObject: []byte(want2),
+			},
+		},
+		{
+			name: "SuccessfulExtractWithDefaulting",
```

Comment:
My intention was to validate the somewhat controversial [upstream behavior](https://github.com/kubernetes/kubernetes/issues/115563), so that we match the upstream behavior. And the aim was, when this behavior somehow changes at later versions of k8s lib, we catch this,and adapt accordingly.
So actually, this test asserts that we extract the defaulted fields too, because this is how the API server records the managedFields for defaulted fields.

Though I realized that to be able to properly test this as I intented, we need to start from an "upper layer" and create an object with SSA, then extract. Because AFAIU, the real cause of the behavior is NOT the extraction logic, but how managed fields are recorded by k8s api server during SSA (that involves defaulting).

Above sounds more like an integration test, let's do this as a follow-up.